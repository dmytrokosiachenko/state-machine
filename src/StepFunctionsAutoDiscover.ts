import * as fs from 'fs';
import * as path from 'path';
import { basename, dirname, extname, join } from 'path';
import { pascal } from 'case';
import { Component, Project, SourceCode } from 'projen';
import { AutoDiscoverCommonOptions } from 'projen/lib/awscdk';
import { AwsCdkDeps } from 'projen/lib/awscdk/awscdk-deps';
import { AutoDiscoverBase } from 'projen/lib/cdk';
import { buildStateType } from './BuildStateType';

const JSON_STEPFUNCTION_EXT = '.workflow.json';

export interface StepFunctionsAutoDiscoverOptions extends AutoDiscoverCommonOptions {
  /**
   * Project source tree (relative to project output directory).
   */
  readonly srcdir: string;
}

export class StepFunctionsAutoDiscover extends AutoDiscoverBase {
  constructor(project: Project, options: StepFunctionsAutoDiscoverOptions) {
    super(project, {
      extension: JSON_STEPFUNCTION_EXT,
      projectdir: options.srcdir,
    });
    for (const entrypoint of this.entrypoints) {
      new StepFunctionsStateMachine(this.project, {
        workflowAsl: entrypoint,
        cdkDeps: options.cdkDeps,
      });
    }
  }
}

export interface StepFunctionsStateMachineOptions {
  readonly constructFile?: string;
  readonly workflowAsl: string;
  readonly constructName?: string;
  readonly cdkDeps: AwsCdkDeps;
}

export class StepFunctionsStateMachine extends Component {
  constructor(project: Project, options: StepFunctionsStateMachineOptions) {
    super(project);

    const workflowAsl = options.workflowAsl;

    if (
      !workflowAsl.endsWith(JSON_STEPFUNCTION_EXT)
    ) {
      throw new Error(
        `${workflowAsl} must have a ${JSON_STEPFUNCTION_EXT} extension`,
      );
    }
    const basePath = join(
      dirname(workflowAsl),
      basename(
        workflowAsl,
        JSON_STEPFUNCTION_EXT,
      ),
    );
    const constructFile = options.constructFile ?? `${basePath}-statemachine.ts`;

    if (extname(constructFile) !== '.ts') {
      throw new Error(
        `Construct file name "${constructFile}" must have a .ts extension`,
      );
    }

    // type names
    const constructName =
      options.constructName ?? pascal(basename(basePath)) + 'StateMachine';

    const src = new SourceCode(project, constructFile);
    if (src.marker) {
      src.line(`// ${src.marker}`);
    }
    src.line('import { StateMachine, StateMachineProps } from \'@matthewbonig/state-machine\';');
    src.line('import { Construct } from \'constructs\';');
    src.line('import fs from \'fs\';');
    src.line('import path from \'path\';');

    src.open(`export interface ${constructName}Overrides {`);

    const workflowDefinition = JSON.parse(fs.readFileSync(workflowAsl).toString());
    if (!workflowDefinition.States) {
      throw new Error(`The workflow file ${workflowAsl} doesn't appear to be a valid ASL file, it doesn't contain a 'States' field`);
    }
    for (const [stateName, stateDef] of Object.entries(workflowDefinition.States)) {
      const stateType = buildStateType(stateDef);
      src.line(`readonly '${stateName}': ${stateType};`);
    }
    src.close('}');

    src.open(`export interface ${constructName}Props extends Omit<StateMachineProps, 'overrides' | 'definition'> {`);
    src.line(`readonly overrides: ${constructName}Overrides;`);
    src.close('}');


    src.open(`export class ${constructName} extends StateMachine {`);
    src.open(`constructor(scope: Construct, id: string, props: ${constructName}Props) {`);
    src.open('super(scope, id, {');
    src.line('...props,');
    const relativeFile = path.basename(workflowAsl);
    src.line(`definition: JSON.parse(fs.readFileSync(path.join(__dirname, '${relativeFile}')).toString()),`);
    src.close('});');
    src.close('}');
    src.close('}');

    this.project.logger.verbose(
      `${basePath}: construct "${constructName}" generated under "${constructFile}"`,
    );
  }
}
