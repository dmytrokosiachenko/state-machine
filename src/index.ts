export * from './StateMachine';
export { StepFunctionsAutoDiscover, StepFunctionsAutoDiscoverOptions } from './StepFunctionsAutoDiscover';
