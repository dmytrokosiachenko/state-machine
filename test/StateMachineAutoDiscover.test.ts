import * as fs from 'fs';
import { mkdtempSync } from 'fs';
import * as os from 'os';
import * as path from 'path';
import { AwsCdkTypeScriptApp } from 'projen/lib/awscdk';
import { synthSnapshot } from 'projen/lib/util/synth';
import { StepFunctionsAutoDiscover } from '../src';


describe('AutoDiscover', () => {

  function setupTestProject(testFile: string) {
    const tempDir = mkdtempSync(path.join(os.tmpdir(), 'test2'));
    const project = new AwsCdkTypeScriptApp({
      name: 'test',
      defaultReleaseBranch: 'main',
      cdkVersion: '2.53.0',
      outdir: tempDir,
    });

    // add a test file...
    fs.mkdirSync(path.join(project.outdir, project.srcdir));
    fs.copyFileSync(
      path.join(__dirname, 'step-functions', testFile),
      path.join(project.outdir, project.srcdir, 'test.workflow.json'),
    );
    new StepFunctionsAutoDiscover(project);
    return project;
  }

  test('test 1', () => {
    const project = setupTestProject('test.workflow.json');
    const snap = synthSnapshot(project);
    expect(snap['src/test-statemachine.ts']).toMatchSnapshot();
  });

  test('test 2', async () => {
    const project = setupTestProject('test2.workflow.json');
    const snap = synthSnapshot(project);
    expect(snap['src/test-statemachine.ts']).toMatchSnapshot();
  });
});
