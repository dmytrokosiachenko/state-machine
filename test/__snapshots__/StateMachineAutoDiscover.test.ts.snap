// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoDiscover test 1 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
import { StateMachine, StateMachineProps } from '@matthewbonig/state-machine';
import { Construct } from 'constructs';
import fs from 'fs';
import path from 'path';
export interface TestStateMachineOverrides {
  readonly 'Which database?': Partial<{
    'Type': string;
    'Choices': [Partial<{
      'Variable': string;
      'StringMatches': string;
      'Next': string;
    }>,Partial<{
      'Variable': string;
      'StringMatches': string;
      'Next': string;
    }>,Partial<{
      'Variable': string;
      'StringMatches': string;
      'Next': string;
    }>];
  }>;
  readonly 'Get belle secret': Partial<{
    'Type': string;
    'Next': string;
    'Parameters': Partial<{
      'SecretId': string;
    }>;
    'Resource': string;
    'ResultPath': string;
  }>;
  readonly 'Parse belle secret value': Partial<{
    'Type': string;
    'Parameters': Partial<{
      'ParsedValue.$': string;
    }>;
    'ResultPath': string;
    'Next': string;
  }>;
  readonly 'Create new database username and password': Partial<{
    'Type': string;
    'Resource': string;
    'Parameters': Partial<{
      'Payload.$': string;
      'FunctionName': string;
    }>;
    'Retry': [Partial<{
      'ErrorEquals': [string,string,string];
      'IntervalSeconds': number;
      'MaxAttempts': number;
      'BackoffRate': number;
    }>];
    'ResultPath': string;
    'Comment': string;
    'Next': string;
    'ResultSelector': Partial<{
      'username.$': string;
      'password.$': string;
      'otsUrl.$': string;
    }>;
  }>;
  readonly 'Create user on the database': Partial<{
    'Type': string;
    'Resource': string;
    'Parameters': Partial<{
      'FunctionName': string;
      'Payload': Partial<{
        'script.$': string;
        'databaseName.$': string;
      }>;
    }>;
    'Retry': [Partial<{
      'ErrorEquals': [string,string,string];
      'IntervalSeconds': number;
      'MaxAttempts': number;
      'BackoffRate': number;
    }>];
    'ResultPath': string;
    'Next': string;
  }>;
  readonly 'Send credentials to user': Partial<{
    'Type': string;
    'Parameters': Partial<{
      'Destination': Partial<{
        'ToAddresses.$': string;
      }>;
      'Message': Partial<{
        'Body': Partial<{
          'Text': Partial<{
            'Data.$': string;
          }>;
        }>;
        'Subject': Partial<{
          'Data': string;
        }>;
      }>;
      'Source': string;
    }>;
    'Resource': string;
    'End': Partial<{
    }>;
  }>;
  readonly 'Get engines secret': Partial<{
    'Type': string;
    'Parameters': Partial<{
      'SecretId': string;
    }>;
    'Resource': string;
    'ResultPath': string;
    'Next': string;
  }>;
  readonly 'Parse engines secret value': Partial<{
    'Type': string;
    'Parameters': Partial<{
      'ParsedValue.$': string;
    }>;
    'ResultPath': string;
    'Next': string;
  }>;
  readonly 'Create new database username and password for engines': Partial<{
    'Type': string;
    'Resource': string;
    'Parameters': Partial<{
      'Payload.$': string;
      'FunctionName': string;
    }>;
    'Retry': [Partial<{
      'ErrorEquals': [string,string,string];
      'IntervalSeconds': number;
      'MaxAttempts': number;
      'BackoffRate': number;
    }>];
    'ResultPath': string;
    'Comment': string;
    'ResultSelector': Partial<{
      'username.$': string;
      'password.$': string;
      'otsUrl.$': string;
    }>;
    'Next': string;
  }>;
  readonly 'Create user on jasmine': Partial<{
    'Type': string;
    'Resource': string;
    'Parameters': Partial<{
      'FunctionName': string;
      'Payload': Partial<{
        'script.$': string;
        'databaseName.$': string;
      }>;
    }>;
    'Retry': [Partial<{
      'ErrorEquals': [string,string,string];
      'IntervalSeconds': number;
      'MaxAttempts': number;
      'BackoffRate': number;
    }>];
    'ResultPath': string;
    'Next': string;
  }>;
  readonly 'Send credentials to jasmine user': Partial<{
    'Type': string;
    'Parameters': Partial<{
      'Destination': Partial<{
        'ToAddresses.$': string;
      }>;
      'Message': Partial<{
        'Body': Partial<{
          'Text': Partial<{
            'Data.$': string;
          }>;
        }>;
        'Subject': Partial<{
          'Data': string;
        }>;
      }>;
      'Source': string;
    }>;
    'Resource': string;
    'End': Partial<{
    }>;
  }>;
  readonly 'Get anna secret': Partial<{
    'Type': string;
    'Next': string;
    'Parameters': Partial<{
      'SecretId': string;
    }>;
    'Resource': string;
    'ResultPath': string;
  }>;
  readonly 'Parse anna secret value': Partial<{
    'Type': string;
    'Parameters': Partial<{
      'ParsedValue.$': string;
    }>;
    'ResultPath': string;
    'Next': string;
  }>;
  readonly 'Create new database username and password for anna': Partial<{
    'Type': string;
    'Resource': string;
    'Parameters': Partial<{
      'Payload.$': string;
      'FunctionName': string;
    }>;
    'Retry': [Partial<{
      'ErrorEquals': [string,string,string];
      'IntervalSeconds': number;
      'MaxAttempts': number;
      'BackoffRate': number;
    }>];
    'ResultPath': string;
    'Comment': string;
    'Next': string;
    'ResultSelector': Partial<{
      'username.$': string;
      'password.$': string;
      'otsUrl.$': string;
    }>;
  }>;
  readonly 'Create user on the anna database': Partial<{
    'Type': string;
    'Resource': string;
    'Parameters': Partial<{
      'FunctionName': string;
      'Payload': Partial<{
        'script.$': string;
        'databaseName.$': string;
      }>;
    }>;
    'Retry': [Partial<{
      'ErrorEquals': [string,string,string];
      'IntervalSeconds': number;
      'MaxAttempts': number;
      'BackoffRate': number;
    }>];
    'ResultPath': string;
    'Next': string;
  }>;
  readonly 'Send credentials to user for anna': Partial<{
    'Type': string;
    'Parameters': Partial<{
      'Destination': Partial<{
        'ToAddresses.$': string;
      }>;
      'Message': Partial<{
        'Body': Partial<{
          'Text': Partial<{
            'Data.$': string;
          }>;
        }>;
        'Subject': Partial<{
          'Data': string;
        }>;
      }>;
      'Source': string;
    }>;
    'Resource': string;
    'End': Partial<{
    }>;
  }>;
}
export interface TestStateMachineProps extends Omit<StateMachineProps, 'overrides' | 'definition'> {
  readonly overrides: TestStateMachineOverrides;
}
export class TestStateMachine extends StateMachine {
  constructor(scope: Construct, id: string, props: TestStateMachineProps) {
    super(scope, id, {
      ...props,
      definition: JSON.parse(fs.readFileSync(path.join(__dirname, 'test.workflow.json')).toString()),
    });
  }
}"
`;

exports[`AutoDiscover test 2 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
import { StateMachine, StateMachineProps } from '@matthewbonig/state-machine';
import { Construct } from 'constructs';
import fs from 'fs';
import path from 'path';
export interface TestStateMachineOverrides {
  readonly 'Get Pod To Execute On': Partial<{
    'Type': string;
    'Resource': string;
    'Parameters': Partial<{
      'FunctionName': string;
      'Payload': Partial<{
        'commands': [string,string,string,string];
      }>;
    }>;
    'Retry': [Partial<{
      'ErrorEquals': [string,string,string];
      'IntervalSeconds': number;
      'MaxAttempts': number;
      'BackoffRate': number;
    }>];
    'ResultSelector': Partial<{
      'belle.$': string;
    }>;
    'ResultPath': string;
    'Next': string;
    'Catch': [Partial<{
      'ErrorEquals': [string];
      'Next': string;
      'ResultPath': string;
    }>];
  }>;
  readonly 'Generate Commands': Partial<{
    'Type': string;
    'Resource': string;
    'Parameters': Partial<{
      'Payload.$': string;
      'FunctionName': string;
    }>;
    'Retry': [Partial<{
      'ErrorEquals': [string,string,string];
      'IntervalSeconds': number;
      'MaxAttempts': number;
      'BackoffRate': number;
    }>];
    'ResultSelector': Partial<{
      'commands.$': string;
    }>;
    'ResultPath': string;
    'Next': string;
    'Catch': [Partial<{
      'ErrorEquals': [string];
      'Next': string;
      'ResultPath': string;
    }>];
  }>;
  readonly 'Run Command': Partial<{
    'Type': string;
    'Resource': string;
    'Parameters': Partial<{
      'Payload': Partial<{
        'commands.$': string;
      }>;
      'FunctionName': string;
    }>;
    'Retry': [Partial<{
      'ErrorEquals': [string,string,string];
      'IntervalSeconds': number;
      'MaxAttempts': number;
      'BackoffRate': number;
    }>];
    'ResultPath': string;
    'Next': string;
    'Catch': [Partial<{
      'ErrorEquals': [string];
      'Next': string;
      'ResultPath': string;
    }>];
  }>;
  readonly 'Notify Slack': Partial<{
    'Type': string;
    'Resource': string;
    'Parameters': Partial<{
      'Payload.$': string;
      'FunctionName': string;
    }>;
    'Retry': [Partial<{
      'ErrorEquals': [string,string,string];
      'IntervalSeconds': number;
      'MaxAttempts': number;
      'BackoffRate': number;
    }>];
    'ResultPath': any;
    'End': Partial<{
    }>;
    'Catch': [Partial<{
      'ErrorEquals': [string];
      'Next': string;
      'ResultPath': string;
    }>];
  }>;
  readonly 'Notify Slack of Error': Partial<{
    'Type': string;
    'Resource': string;
    'Parameters': Partial<{
      'Payload.$': string;
      'FunctionName': string;
    }>;
    'Retry': [Partial<{
      'ErrorEquals': [string,string,string];
      'IntervalSeconds': number;
      'MaxAttempts': number;
      'BackoffRate': number;
    }>];
    'End': Partial<{
    }>;
    'ResultPath': any;
  }>;
}
export interface TestStateMachineProps extends Omit<StateMachineProps, 'overrides' | 'definition'> {
  readonly overrides: TestStateMachineOverrides;
}
export class TestStateMachine extends StateMachine {
  constructor(scope: Construct, id: string, props: TestStateMachineProps) {
    super(scope, id, {
      ...props,
      definition: JSON.parse(fs.readFileSync(path.join(__dirname, 'test.workflow.json')).toString()),
    });
  }
}"
`;
